app/routes.py

```diff
@@ -1 +1,280 @@
-from flask import Blueprint
\ No newline at end of file
+import os
+from datetime import datetime
+
+import requests
+from flask import Blueprint, abort, jsonify, make_response, request
+
+from app import db
+from app.models.goal import Goal
+from app.models.task import Task
+
+task_bp = Blueprint("tasks", __name__, url_prefix="/tasks")
+goal_bp = Blueprint("goals", __name__, url_prefix="/goals")
+
+##-----------------------------------------------------------------##
+##--------------------------Helper Functions-----------------------##
+##-----------------------------------------------------------------##
+
+def validate_item(cls,id):
+    try:
+        id = int(id)
+    except:
+        abort(make_response({"message": f"{cls.__name__} {id} is invalid"}, 400))
+
+    item = cls.query.get(id)
+    if not item:
+        abort(make_response({"message": f"{cls.__name__} {id} not found"}, 404))
+
+    return item
+
+def validate_new_item(cls,request_body):
+    try:  
+        if cls == Task:
+            new_item = cls(title=request_body["title"],
+                           description=request_body["description"]
+                        )
+        else:
+             new_item = cls(title=request_body["title"])
+    except:
+        abort(make_response({"details": "Invalid data"},400))
+
+    return new_item
+
+def post_to_slack(task):
+    URL = "https://slack.com/api/chat.postMessage"
+    token = os.environ.get('SLACK_TOKEN')
+    Headers = {"Authorization" : f'Bearer {token}'}
+    params = {'channel':'task-notifications',
+              'text': f'Someone just completed the task {task.title}'
+            }
+
+    requests.post(URL,headers=Headers,params=params)
+
+def task_to_goal(task_id,goal):
+    task = validate_item(Task,task_id)
+    task.goal_id=goal.goal_id
+
+    return task_id
+
+def to_dict(self,list_tasks=False):
+    item_dict = {} 
+
+    if self.__class__ == Task:
+        item_dict["id"]=self.task_id
+        item_dict["title"]=self.title
+        item_dict["description"]=self.description
+
+        if self.completed_at != None:
+            item_dict["is_complete"]=True
+        else:
+            item_dict["is_complete"]=False
+
+        if self.goal_id:
+            item_dict["goal_id"]=self.goal_id
+        
+    if self.__class__ == Goal:
+        item_dict["id"]=self.goal_id
+        item_dict["title"]=self.title
+
+        if list_tasks==True:
+            task_list = []
+            for task in self.tasks:
+                task_list.append(to_dict(task)) # recursion baby!
+            item_dict["tasks"]=task_list
+
+    return item_dict
+
+def query_filter(cls):
+    sort_query = request.args.get("sort")
+    filter_query = request.args.get("filter")
+    by = request.args.get("by")
+
+    if sort_query:
+        if sort_query == 'asc' or sort_query != 'desc':
```

Comment:
This handles a lot of cases, which is nice! It would be even better to be able to build up the SQL Alchemy query without needing all of the conditional logic. Using a lot of conditional logic can be hard to maintain and sometimes challenging to read. Also the logic complexity would increase substantially if we added more ways of filtering the query.